main/src/ca/uwaterloo/flix/language/phase/Uncurrier2.scala

```diff
@@ -0,0 +1,366 @@
+          val rs = rules map {
+            case SelectChannelRule(sym, chan, exp) =>
+              val c = uncurryApply(chan)
+              val e = uncurryApply(exp)
+              SelectChannelRule(sym, c, e)
+          }
+          val d = default.map(uncurryApply)
+          Expression.SelectChannel(rs, d, tpe, loc)
+
+        case Expression.ProcessSpawn(exp, tpe, loc) =>
+          val e = uncurryApply(exp)
+          Expression.ProcessSpawn(e, tpe, loc)
+
+        case Expression.ProcessPanic(msg, tpe, loc) => Expression.ProcessPanic(msg, tpe, loc)
+
+        case Expression.FixpointConstraintSet(cs0, tpe, loc) => Expression.FixpointConstraintSet(cs0, tpe, loc)
+
+        case Expression.FixpointCompose(exp1, exp2, tpe, loc) =>
+          val e1 = uncurryApply(exp1)
+          val e2 = uncurryApply(exp2)
+          Expression.FixpointCompose(e1, e2, tpe, loc)
+
+        case Expression.FixpointSolve(exp, stf, tpe, loc) =>
+          val e = uncurryApply(exp)
+          Expression.FixpointSolve(e, stf, tpe, loc)
+
+        case Expression.FixpointProject(sym, exp, tpe, loc) =>
+          val e = uncurryApply(exp)
+          Expression.FixpointProject(sym, e, tpe, loc)
+
+        case Expression.FixpointEntails(exp1, exp2, tpe, loc) =>
+          val e1 = uncurryApply(exp1)
+          val e2 = uncurryApply(exp2)
+          Expression.FixpointEntails(e1, e2, tpe, loc)
+
+        case Expression.FixpointFold(sym, exp1, exp2, exp3, tpe, loc) =>
+          val e1 = uncurryApply(exp1)
+          val e2 = uncurryApply(exp2)
+          val e3 = uncurryApply(exp3)
+          Expression.FixpointFold(sym, e1, e2, e3, tpe, loc)
+
+        case Expression.HoleError(sym, tpe, loc) => Expression.HoleError(sym, tpe, loc)
+        case Expression.MatchError(tpe, loc) => Expression.MatchError(tpe, loc)
+
+        case Expression.LambdaClosure(fparams, freeVars, exp, tpe, loc) => throw InternalCompilerException(s"Unexpected expression: '${exp0.getClass}'.")
+        case Expression.Closure(sym, freeVars, tpe, loc) => throw InternalCompilerException(s"Unexpected expression: '${exp0.getClass}'.")
+        case Expression.ApplyClo(exp, args, tpe, loc) => throw InternalCompilerException(s"Unexpected expression: '${exp0.getClass}'.")
+        case Expression.ApplyDef(sym, args, tpe, loc) => throw InternalCompilerException(s"Unexpected expression: '${exp0.getClass}'.")
+        case Expression.ApplyCloTail(exp, args, tpe, loc) => throw InternalCompilerException(s"Unexpected expression: '${exp0.getClass}'.")
+        case Expression.ApplyDefTail(sym, args, tpe, loc) => throw InternalCompilerException(s"Unexpected expression: '${exp0.getClass}'.")
+        case Expression.ApplySelfTail(sym, formals, actuals, tpe, loc) => throw InternalCompilerException(s"Unexpected expression: '${exp0.getClass}'.")
+      }
+
+    val uncurriedDefs1 = flix.subphase("Uncurrying step 1") {
+      uncurriedDefs0 map {
```

Comment:
This could be done in parallel with parMap?

Reply:
Yes, it's just a local rewrite on each definition and the shared map is only used for lookups. But parMap works on lists and the program is represented as a map. There is no speedup, [in fact a slowdown](https://puu.sh/FuM0y/4fae3b5da0.png). I assume the degree of parallelism on my machine is not enough to make up for the overhead of converting uncurriedDefs0 to a list and then uncurriedDefs1 into a map.

Reply:
Interesting. Let us talk about this tomorrow. Might be a fun little experiment to include in the report.